In [1]:
from ete2 import NCBITaxa
from Bio import Entrez
from collections import OrderedDict
import pandas as pd
import json
import re

In [2]:
ncbi = NCBITaxa()

Update taxonomy database, might takes a few minutes...

In [3]:
#ncbi.update_taxonomy_database()

#### Insert the root of the taxonomy to start retrieving information from

In [4]:
organism = "Teuthida"

In [5]:
taxid2name = ncbi.get_name_translator([organism])
taxid2name

{'Teuthida': [551290]}

In [6]:
organism_taxid = taxid2name[organism][0]
organism_taxid

551290

Available methods
- NCBITaxa.get_rank()
- NCBITaxa.get_lineage()
- NCBITaxa.get_taxid_translator()
- NCBITaxa.get_name_translator()
- NCBITaxa.translate_to_names()

In [7]:
descendants = ncbi.get_descendant_taxa(organism, intermediate_nodes=True)
print ncbi.translate_to_names(descendants[:10])

[u'Loligo sp.', u'environmental samples', u'Dosidicus gigas environmental sample', u'Loliolus', u'Teuthida sp. BOLD:AAB0336', u'Teuthida sp. BOLD:AAD8239', u'Loliolus uyii', u'Sthenoteuthis sp. 1 SS-2016', u'Sthenoteuthis sp. 2 SS-2016', u'Sthenoteuthis sp. 3 SS-2016']


In [8]:
print("Ci sono {} nodi nella tassonomia dei {}".format(len(descendants), organism))

Ci sono 614 nodi nella tassonomia dei Teuthida


In [9]:
ancestor_ranks = ncbi.get_lineage(organism_taxid)
ancestor_ranks

[1,
 131567,
 2759,
 33154,
 33208,
 6072,
 33213,
 33317,
 1206795,
 6447,
 6605,
 6606,
 215449,
 215450,
 551290]

In [10]:
full_ranks = ncbi.get_rank(ancestor_ranks + descendants)
full_ranks[1] = u'root' # if not it is 'no rank'

#ranks = ncbi.get_rank(descendants)
ranks = ncbi.get_rank(descendants + [organism_taxid])# include self

full_ranks

{1: u'root',
 2759: u'superkingdom',
 6072: u'no rank',
 6447: u'phylum',
 6605: u'class',
 6606: u'subclass',
 6615: u'family',
 6616: u'genus',
 6618: u'species',
 6622: u'species',
 6623: u'family',
 6624: u'genus',
 6625: u'species',
 6626: u'family',
 6627: u'genus',
 6628: u'species',
 6629: u'species',
 6630: u'genus',
 6631: u'species',
 6632: u'genus',
 6636: u'genus',
 6637: u'species',
 33154: u'no rank',
 33208: u'kingdom',
 33213: u'no rank',
 33317: u'no rank',
 34542: u'suborder',
 34543: u'genus',
 34544: u'species',
 34545: u'genus',
 34546: u'species',
 34547: u'genus',
 34548: u'species',
 34549: u'genus',
 34550: u'species',
 34551: u'species',
 34552: u'genus',
 34553: u'species',
 34554: u'family',
 34555: u'genus',
 34556: u'species',
 34557: u'family',
 34558: u'genus',
 34559: u'species',
 34560: u'family',
 34561: u'genus',
 34562: u'species',
 34563: u'genus',
 34564: u'species',
 34565: u'family',
 34566: u'genus',
 34567: u'species',
 34569: u'genus',
 3457

#### Build dictionary of taxid and the its name
e.g. {551290: u'Teuthida'}

In [11]:
taxid_translator = {}
for taxid in full_ranks:
    taxid_translator[taxid] = ncbi.get_taxid_translator([taxid])[taxid]
taxid_translator

{1: u'root',
 2759: u'Eukaryota',
 6072: u'Eumetazoa',
 6447: u'Mollusca',
 6605: u'Cephalopoda',
 6606: u'Coleoidea',
 6615: u'Loliginidae',
 6616: u'Loligo',
 6618: u'Loligo forbesii',
 6622: u'Loligo vulgaris',
 6623: u'Enoploteuthidae',
 6624: u'Watasenia',
 6625: u'Watasenia scintillans',
 6626: u'Ommastrephidae',
 6627: u'Illex',
 6628: u'Illex argentinus',
 6629: u'Illex illecebrosus',
 6630: u'Nototodarus',
 6631: u'Nototodarus gouldi',
 6632: u'Ommastrephes',
 6636: u'Todarodes',
 6637: u'Todarodes pacificus',
 33154: u'Opisthokonta',
 33208: u'Metazoa',
 33213: u'Bilateria',
 33317: u'Protostomia',
 34542: u'Oegopsina',
 34543: u'Pterygioteuthis',
 34544: u'Pterygioteuthis microlampas',
 34545: u'Enoploteuthis',
 34546: u'Enoploteuthis reticulata',
 34547: u'Abraliopsis',
 34548: u'Abraliopsis sp.',
 34549: u'Abralia',
 34550: u'Abralia trigonura',
 34551: u'Nototodarus hawaiiensis',
 34552: u'Sthenoteuthis',
 34553: u'Sthenoteuthis oualaniensis',
 34554: u'Architeuthidae',
 

#### Create a dictionary of lineage for each taxid we fetched
It is structured in the following way: <br>
TAXID: {GRUPPO_TASSONOMICO: ISTANZA} for each rank<br>
e.g. 
  * {551290: 
    * {..., 
        * 'infraclass': 'Neocoleoidea'},
            * {'superorder', 'Decapodiformes'},
                * {order: Teuthida}}

In [12]:
# taxid_lineage = {}
# for taxid, rank in ranks.items():
#     taxid_lineage[taxid] = OrderedDict()
#     count_noranks = 0
#     for ancestor_id in ncbi.get_lineage(key):
#         lineage_level_name = taxid_translator[ancestor_id]
#         lineage_instance = full_ranks[ancestor_id] # u'Teuthida, u'Cephalopoda, etc...
#         # do not override no rank keys !
#         if lineage_instance == u'no rank':
#             if count_noranks > 0:
#                 lineage_instance += '_{}'.format(str(count_noranks)) #no rank_1
#             count_noranks += 1
#         taxid_lineage[taxid][lineage_instance] = lineage_level_name # set e.g. {u'superkingdom: u'Eukaryota'}
# taxid_lineage

In [13]:
taxid_lineage = {}
for taxid, rank in ranks.items():
    taxid_lineage[taxid] = OrderedDict()
    count_noranks = 0
    count_consecutive_noranks = 1 # e' il primo no rank consecutivo --> e' la prima volta che occorre
    was_norank = False
    for i, ancestor_id in enumerate(ncbi.get_lineage(taxid)):
        lineage_level_name = taxid_translator[ancestor_id] # u'Teuthida, u'Cephalopoda, etc...
        lineage_instance = full_ranks[ancestor_id] # order, suborder, ...
        # do not override no rank keys !
        if lineage_instance == u'no rank': # first instance is never no rank, else code will crash
            # if the previous ancestor is not on the same level then reset counter 
            if not was_norank:
                lineage_instance = u'sub_' + taxid_lineage[taxid].items()[i-1][0] # take the upper ancestor
            else:
                # take the upper common ancestor
                lineage_instance = u'sub_{}_{}'.format(taxid_lineage[taxid].items()[i-1-count_consecutive_noranks][0],
                                                        count_consecutive_noranks)
                count_consecutive_noranks += 1
            was_norank = True
        else:
            count_consecutive_noranks = 1
            was_norank = False
        taxid_lineage[taxid][lineage_instance] = lineage_level_name # set e.g. {u'superkingdom: u'Eukaryota'}
taxid_lineage

# PSEUDOCODICE
#count_consecutive_noranks = 1
#per ogni ancestor
    #se ho un no_rank
        #se was_norank == False
            #chiamalo sotto_livello
        #altrimenti
            #chiamalo sotto_livello_${count_consecutive_noranks}
            #count_consecutive_noranks += 1
        #was_norank = True
    #se non ho un no_rank
        #was_norank = False

{6615: OrderedDict([(u'root', u'root'),
              (u'sub_root', u'cellular organisms'),
              (u'superkingdom', u'Eukaryota'),
              (u'sub_superkingdom', u'Opisthokonta'),
              (u'kingdom', u'Metazoa'),
              (u'sub_kingdom', u'Eumetazoa'),
              (u'sub_kingdom_1', u'Bilateria'),
              (u'sub_kingdom_2', u'Protostomia'),
              (u'sub_kingdom_3', u'Lophotrochozoa'),
              (u'phylum', u'Mollusca'),
              (u'class', u'Cephalopoda'),
              (u'subclass', u'Coleoidea'),
              (u'infraclass', u'Neocoleoidea'),
              (u'superorder', u'Decapodiformes'),
              (u'order', u'Teuthida'),
              (u'suborder', u'Myopsina'),
              (u'family', u'Loliginidae')]),
 6616: OrderedDict([(u'root', u'root'),
              (u'sub_root', u'cellular organisms'),
              (u'superkingdom', u'Eukaryota'),
              (u'sub_superkingdom', u'Opisthokonta'),
              (u'kingdom', u

#### Manual part: reorder columns (last columns are the ones that were not present in the first few taxids (greedy deep-first))

In [14]:
columns_set = ()
for k, v in taxid_lineage.items():
    for k1 in v:
        if k1 not in columns_set:
            columns_set = columns_set + (k1,)
# Order is correct until 'order' :)
idx_untilcorrect = columns_set.index("order")
columns_set

(u'root',
 u'sub_root',
 u'superkingdom',
 u'sub_superkingdom',
 u'kingdom',
 u'sub_kingdom',
 u'sub_kingdom_1',
 u'sub_kingdom_2',
 u'sub_kingdom_3',
 u'phylum',
 u'class',
 u'subclass',
 u'infraclass',
 u'superorder',
 u'order',
 u'suborder',
 u'family',
 u'genus',
 u'sub_genus',
 u'species',
 u'sub_family',
 u'sub_suborder',
 u'subspecies',
 u'sub_order')

In [15]:
columns = list(columns_set[:idx_untilcorrect + 1])
columns.extend([u'sub_order', 
                u'suborder', 
                u'sub_suborder', 
                u'family', 
                u'sub_family',
                u'genus',
                u'sub_genus',
                u'species', 
                u'subspecies'])

print("Did I miss any column? {}".format(len(columns) != len(columns_set)))

Did I miss any column? False


In [16]:
dd = pd.DataFrame.from_dict(taxid_lineage, orient = "index", columns=columns)
dd.iloc[:,-10:]

order sub_order   suborder sub_suborder            family  \
6615     Teuthida       NaN   Myopsina          NaN       Loliginidae   
6616     Teuthida       NaN   Myopsina          NaN       Loliginidae   
6618     Teuthida       NaN   Myopsina          NaN       Loliginidae   
6622     Teuthida       NaN   Myopsina          NaN       Loliginidae   
6623     Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
6624     Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
6625     Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
6626     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6627     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6628     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6629     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6630     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6631     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6632     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6636     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
6637     Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
34542    Teuthida       NaN  Oegopsina          NaN               NaN   
34543    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34544    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34545    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34546    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34547    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34548    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34549    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34550    Teuthida       NaN  Oegopsina          NaN   Enoploteuthidae   
34551    Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
34552    Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
34553    Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
34554    Teuthida       NaN  Oegopsina          NaN    Architeuthidae   
34555    Teuthida       NaN  Oegopsina          NaN    Architeuthidae   
...           ...       ...        ...          ...               ...   
2060711  Teuthida       NaN  Oegopsina          NaN  Mastigoteuthidae   
2094581  Teuthida       NaN  Oegopsina          NaN    Ommastrephidae   
2163882  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2163883  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306799  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306800  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306801  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306802  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306803  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306804  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306805  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306806  Teuthida       NaN  Oegopsina          NaN   Walvisteuthidae   
2306807  Teuthida       NaN  Oegopsina          NaN   Walvisteuthidae   
2306808  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306809  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306810  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306811  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306812  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306818  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306960  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306961  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306963  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae   
2306964  Teuthida       NaN  Oegopsina          NaN   Onychoteuthidae 

In [17]:
filter_col = [col for col in dd if col.startswith('sub_')]
for f in filter_col:
    print("Column '{}' has {} unique value(s): {}".format(f, len(dd[f].unique()), dd[f].unique()))

Column 'sub_root' has 1 unique value(s): [u'cellular organisms']
Column 'sub_superkingdom' has 1 unique value(s): [u'Opisthokonta']
Column 'sub_kingdom' has 1 unique value(s): [u'Eumetazoa']
Column 'sub_kingdom_1' has 1 unique value(s): [u'Bilateria']
Column 'sub_kingdom_2' has 1 unique value(s): [u'Protostomia']
Column 'sub_kingdom_3' has 1 unique value(s): [u'Lophotrochozoa']
Column 'sub_order' has 3 unique value(s): [nan u'unclassified Teuthida' u'environmental samples']
Column 'sub_suborder' has 3 unique value(s): [nan u'unclassified Myopsina' u'environmental samples']
Column 'sub_family' has 3 unique value(s): [nan u'environmental samples' u'unclassified Cranchiidae']
Column 'sub_genus' has 4 unique value(s): [nan u'unclassified Loligo' u'unclassified Illex' u'environmental samples']


In [18]:
# select no_rank columns rooting (starting from) at the chosen organism i.e. avoid ancestors' no ranks
try:
    idx_filter = filter_col.index("sub_" + ncbi.get_rank([organism_taxid])[organism_taxid])
except:
    idx_filter = 0
norank_col = filter_col[idx_filter:]
norank_col

[u'sub_order', u'sub_suborder', u'sub_family', u'sub_genus']

In [19]:
# dataframe with only those organism that have at least one no rank in the lineage
norank_df = dd[dd[norank_col].notnull().any(axis = 1)]

### Create dataset of taxid and related information

In [20]:
# First build a dictionary...
df = {}
for taxid in descendants + [organism_taxid]:
    df[taxid] = {}
    
    specie = ncbi.translate_to_names([taxid])
    rank_dict = ncbi.get_rank([taxid])
    lineage_id = ncbi.get_lineage(taxid)
    names = ncbi.get_taxid_translator(lineage_id)
    lineage_name = [names[taxid] for taxid in lineage_id]
    
    df[taxid]['name'] = specie[0]
    df[taxid]['rank'] = rank_dict[taxid]
    df[taxid]['lineage_id'] = '//'.join([str(char) for char in lineage_id])
    df[taxid]['lineage_name'] = '//'.join(lineage_name)
#    df[taxid]['lineage_complete'] = taxid_lineage[taxid]

In [21]:
print(json.dumps(df, indent = 2))

{
  "1091609": {
    "lineage_id": "1//131567//2759//33154//33208//6072//33213//33317//1206795//6447//6605//6606//215449//215450//551290//34542//6626//346248//1091609", 
    "name": "environmental samples", 
    "rank": "no rank", 
    "lineage_name": "root//cellular organisms//Eukaryota//Opisthokonta//Metazoa//Eumetazoa//Bilateria//Protostomia//Lophotrochozoa//Mollusca//Cephalopoda//Coleoidea//Neocoleoidea//Decapodiformes//Teuthida//Oegopsina//Ommastrephidae//Dosidicus//environmental samples"
  }, 
  "1091610": {
    "lineage_id": "1//131567//2759//33154//33208//6072//33213//33317//1206795//6447//6605//6606//215449//215450//551290//34542//6626//346248//1091609//1091610", 
    "name": "Dosidicus gigas environmental sample", 
    "rank": "species", 
    "lineage_name": "root//cellular organisms//Eukaryota//Opisthokonta//Metazoa//Eumetazoa//Bilateria//Protostomia//Lophotrochozoa//Mollusca//Cephalopoda//Coleoidea//Neocoleoidea//Decapodiformes//Teuthida//Oegopsina//Ommastrephidae//Dosidicu

In [22]:
# ... then convert the dictionary to dataframe
data = pd.DataFrame.from_dict(data=df, orient="index")
print(data.head())

                                             lineage_id             name  \
6615  1//131567//2759//33154//33208//6072//33213//33...      Loliginidae   
6616  1//131567//2759//33154//33208//6072//33213//33...           Loligo   
6618  1//131567//2759//33154//33208//6072//33213//33...  Loligo forbesii   
6622  1//131567//2759//33154//33208//6072//33213//33...  Loligo vulgaris   
6623  1//131567//2759//33154//33208//6072//33213//33...  Enoploteuthidae   

         rank                                       lineage_name  
6615   family  root//cellular organisms//Eukaryota//Opisthoko...  
6616    genus  root//cellular organisms//Eukaryota//Opisthoko...  
6618  species  root//cellular organisms//Eukaryota//Opisthoko...  
6622  species  root//cellular organisms//Eukaryota//Opisthoko...  
6623   family  root//cellular organisms//Eukaryota//Opisthoko...  


#### Add ancestor relationship

In [23]:
data['sonof_id'] = None
data['sonof_name'] = None
for index, row in data.iterrows():
    row['sonof_id'] = row['lineage_id'].split('//')[-2] # take father node
    row['sonof_name'] = row['lineage_name'].split('//')[-2] # take father node    
    #row['son_of_(rank_name)'] = data[index, 'son_of']

# Reorder columns
data = data[['name', 'rank', 'sonof_name', 'sonof_id', 'lineage_name', 'lineage_id']] # reorder columns
data.sort_values(by=['lineage_id'], inplace=True) # order rows by lineage id
data.head()

name     rank             sonof_name  \
551290                           Teuthida    order         Decapodiformes   
1986411             environmental samples  no rank               Teuthida   
1986412     Teuthida environmental sample  species  environmental samples   
2002786  Doryteuthis environmental sample  species  environmental samples   
2002799        Illex environmental sample  species  environmental samples   

        sonof_id                                       lineage_name  \
551290    215450  root//cellular organisms//Eukaryota//Opisthoko...   
1986411   551290  root//cellular organisms//Eukaryota//Opisthoko...   
1986412  1986411  root//cellular organisms//Eukaryota//Opisthoko...   
2002786  1986411  root//cellular organisms//Eukaryota//Opisthoko...   
2002799  1986411  root//cellular organisms//Eukaryota//Opisthoko...   

                                                lineage_id  
551290   1//131567//2759//33154//33208//6072//33213//33...  
1986411  1//131567//2759//33154//33208//6072//33213//33...  
1986412  1//131567//2759//33154//33208//6072//33213//33...  
2002786  1//131567//2759//33154//33208//6072//33213//33...  
2002799  1//131567//2759//33154//33208//6072//33213//33...

#### Create dataframe for full taxonomy (including everything)

In [24]:
full_taxonomy = data.join(dd)
full_taxonomy

name        rank  \
551290                                            Teuthida       order   
1986411                              environmental samples     no rank   
1986412                      Teuthida environmental sample     species   
2002786                   Doryteuthis environmental sample     species   
2002799                         Illex environmental sample     species   
34542                                            Oegopsina    suborder   
283043                                       Batoteuthidae      family   
283044                                         Batoteuthis       genus   
283045                                 Batoteuthis skolops     species   
34554                                       Architeuthidae      family   
1986286                              environmental samples     no rank   
1986287  Architeuthidae environmental sample cf. Archit...  subspecies   
1986288  Architeuthidae environmental sample cf. Archit...     species   
34555                                         Architeuthis       genus   
1235433                          Architeuthis sp. NSMT 001     species   
215441                            Architeuthis sp. SJ-2002     species   
248766                           Architeuthis sp. BED-2003     species   
256136                                    Architeuthis dux     species   
279107                           Architeuthis sp. JMS-2004     species   
34556                                     Architeuthis sp.     species   
559516                           Architeuthis sp. ARL-2008     species   
664327                            Architeuthis sp. ML-2009     species   
34557                                      Histioteuthidae      family   
1365855                              environmental samples     no rank   
1365856  Histioteuthidae environmental sample cf. Histo...     species   
1986382  Histioteuthidae environmental sample cf. Histi...     species   
1986408  Histioteuthidae environmental sample cf. Stigm...     species   
1986407                                    Stigmatoteuthis       genus   
2053936                            Stigmatoteuthis arcturi     species   
34558                                        Histioteuthis       genus   
...                                                    ...         ...   
1331656                             Uroteuthis sp. SC-2013     species   
1986004                                Uroteuthis sp. Ba03     species   
55720                                    Uroteuthis edulis     species   
78424                                 Uroteuthis chinensis     species   
78425                                Uroteuthis duvaucelii     species   
78426                                Uroteuthis etheridgei     species   
78427                                 Uroteuthis noctiluca     species   
948796                                  Uroteuthis sibogae     species   
6616                                                Loligo       genus   
184220                                 unclassified Loligo     no rank   
184221                         Loligo sp. AL9407-Loligo-1K     species   
55716                                           Loligo sp.     species   
283042                                    Loligo formosana     species   
55283                                     Loligo reynaudii     species   
6618                                       Loligo forbesii     species   
6622                                       Loligo vulgaris     species   
936531                                  Loligo sp. KW-2010     species   
78434                                         Lolliguncula       genus   
1423690                            Lolliguncula panamensis     species   
559555                              Lolliguncula diomedeae     species   
78435                                  Lolliguncula brevis     species   
948964                               unclassified Myopsina     no rank   
975677                           Myopsida sp. BOLD:AAB3728

#### Create taxonomy for organism that have at least a no_rank level associated

In [25]:
norank_taxonomy = data.join(norank_df, how='right')
norank_taxonomy

name        rank  \
55716                                           Loligo sp.     species   
184220                                 unclassified Loligo     no rank   
184221                         Loligo sp. AL9407-Loligo-1K     species   
184222                                  unclassified Illex     no rank   
184223                           Illex sp. AL9407-Illex-2K     species   
517129                               environmental samples     no rank   
517130                   Berryteuthis environmental sample     species   
665457                               environmental samples     no rank   
665458                   Sepioteuthis environmental sample     species   
725814                               unclassified Teuthida     no rank   
855049                           Teuthida sp. BOLD:AAB0336     species   
855050                           Teuthida sp. BOLD:AAD8239     species   
948964                               unclassified Myopsina     no rank   
975677                           Myopsida sp. BOLD:AAB3728     species   
978687                           Myopsida sp. BOLD:AAM1743     species   
978688                           Myopsida sp. BOLD:AAM2097     species   
1091609                              environmental samples     no rank   
1091610               Dosidicus gigas environmental sample     species   
1365855                              environmental samples     no rank   
1365856  Histioteuthidae environmental sample cf. Histo...     species   
1365859                              environmental samples     no rank   
1405369                           unclassified Cranchiidae     no rank   
1405370                       Cranchiidae sp. BOLD:ABA4569     species   
1479259                          Teuthida sp. BOLD:ACJ3715     species   
1844018                               Teuthida sp. AS-2016     species   
1986286                              environmental samples     no rank   
1986287  Architeuthidae environmental sample cf. Archit...  subspecies   
1986288  Architeuthidae environmental sample cf. Archit...     species   
1986364                              environmental samples     no rank   
1986366                Chiroteuthidae environmental sample     species   
...                                                    ...         ...   
1986374  Cycloteuthidae environmental sample cf. Cyclot...     species   
1986376  Chiroteuthidae environmental sample cf. Grimal...     species   
1986377  Cranchiidae environmental sample cf. Helicocra...     species   
1986382  Histioteuthidae environmental sample cf. Histi...     species   
1986383                              environmental samples     no rank   
1986384  Mastigoteuthidae environmental sample cf. Idio...     species   
1986385  Mastigoteuthidae environmental sample cf. Idio...     species   
1986386                              environmental samples     no rank   
1986388  Joubiniteuthidae environmental sample cf. Joub...     species   
1986389   Cranchiidae environmental sample cf. Leachia sp.     species   
1986390                              environmental samples     no rank   
1986391  Lepidoteuthidae environmental sample cf. Lepid...     species   
1986395                              environmental samples     no rank   
1986396  Bathyteuthidae environmental sample cf. Bathyt...     species   
1986398                              environmental samples     no rank   
1986400  Octopoteuthidae environmental sample cf. Octop...     species   
1986401  Ommastrephes environmental sample cf. Ommastre...     species   
1986404                              environmental samples     no rank   
1986405  Onychoteuthidae environmental sample cf. Onyki...     species   
1986408  Histioteuthidae environmental sample cf. Stigm...     species   
1986409  Octopoteuthidae environmental sample cf. Tanin...     species   
1986410  Cranchiidae environmental sample cf. Taonius pavo     species   
1986411                              environmental samples

#### Create complete taxonomy (that is the difference between full and no_ranks)

In [32]:
complete_taxonomy = full_taxonomy.loc[full_taxonomy.index.difference(norank_taxonomy.index)]
complete_taxonomy.dropna(axis=1, how = 'all', inplace=True) # remove now columns with all null
complete_taxonomy

name        rank  \
6615                              Loliginidae      family   
6616                                   Loligo       genus   
6618                          Loligo forbesii     species   
6622                          Loligo vulgaris     species   
6623                          Enoploteuthidae      family   
6624                                Watasenia       genus   
6625                    Watasenia scintillans     species   
6626                           Ommastrephidae      family   
6627                                    Illex       genus   
6628                         Illex argentinus     species   
6629                       Illex illecebrosus     species   
6630                              Nototodarus       genus   
6631                       Nototodarus gouldi     species   
6632                             Ommastrephes       genus   
6636                                Todarodes       genus   
6637                      Todarodes pacificus     species   
34542                               Oegopsina    suborder   
34543                         Pterygioteuthis       genus   
34544             Pterygioteuthis microlampas     species   
34545                           Enoploteuthis       genus   
34546                Enoploteuthis reticulata     species   
34547                             Abraliopsis       genus   
34548                         Abraliopsis sp.     species   
34549                                 Abralia       genus   
34550                       Abralia trigonura     species   
34551                 Nototodarus hawaiiensis     species   
34552                           Sthenoteuthis       genus   
34553              Sthenoteuthis oualaniensis     species   
34554                          Architeuthidae      family   
34555                            Architeuthis       genus   
...                                       ...         ...   
2060711                    Mastigopsis hjorti     species   
2094581                   Ommastrephes sp. E3     species   
2163882     Onychoteuthis cf. banksii AL-2018     species   
2163883                 Onychoteuthis sp. AL2     species   
2306799                 Onychoteuthis lacrima     species   
2306800         Onychoteuthis meridiopacifica     species   
2306801            Onychoteuthis sp. HEB-2018     species   
2306802                   Onychoteuthis sp_AL     species   
2306803          Onychoteuthis sp_AL HEB-2018  subspecies   
2306804         Onykia cf. carriboea HEB-2018     species   
2306805           Onykia cf. robsoni HEB-2018     species   
2306806  Walvisteuthis cf. rancureli HEB-2018     species   
2306807          Walvisteuthis sp. C HEB-2018     species   
2306808  Onychoteuthis aff. compacta HEB-2018     species   
2306809    Onychoteuthis cf. banksii HEB-2018     species   
2306810     Onychoteuthis cf. bergii HEB-2018     species   
2306811    Onychoteuthis cf. prolata HEB-2018     species   
2306812          Onykia aff. robsoni HEB-2018     species   
2306818   Moroteuthopsis aff. ingens HEB-2018     species   
2306960              Moroteuthopsis longimana     species   
2306961                 Moroteuthopsis ingens     species   
2306963                   Notonykia africanae     species   
2306964                      Notonykia nesisi     species   
2306965              Onychoteuthis aequimanus     species   
2502174            Brachioteuthis sp. HJ-2019     species   
2502175             Octopoteuthis sp. HJ-2019     species   
2502176     Onychoteuthis cf. banksii HJ-2019     species   
2502177          Helicocranchia sp. A HJ-2019     species   
2502178               Liguriella podophthalma     species   
2516895            Chtenopteryx sp. C CI-2019     species   

                  sonof_name sonof_id  \
6615                Myopsina   551347   
6616             Loliginidae     6615   
6618                  Loligo     6616   
6622                  Loligo     6616   
6623               Oegopsina    34542   
6624        

### Save all dataframes

In [28]:
full_taxonomy.to_csv("taxonomy_full.csv", index_label = 'taxid')
norank_taxonomy.to_csv("taxonomy_norank.csv", index_label = 'taxid')
complete_taxonomy.to_csv("taxonomy_complete.csv", index_label = 'taxid')

#### Remove lineage common to all entries (i.e. until Teuthida included)

In [ ]:
#common_lineage_to_remove = r"root//.*//" + organism
#data.replace(to_replace = common_lineage_to_remove,
#             value = "", inplace = True, regex = True)
#data.head()

In [ ]:
# Manually add Theuthida order - NOT needed anymore, it is already present
# organism_taxid = 551290
# data.loc[organism_taxid] = ["Teuthida", "order", 
#                             "Teuthida", # lineage_name
#                             "1//131567//2759//33154//33208//6072//33213//33317//1206795//6447//6605//6606//215449//215450//551290",
#                            None, None]
# data.tail()

In [ ]:
list_of_series = [data.index, data.index]
df = pd.DataFrame(list_of_series, columns=cols)

Create dataframe of lineage of taxonomy ranks for each taxid<br>That is (taxid:"279107", rank_lineage: "order//suborder//family//genus//species")

In [ ]:
id_taxidLineage = data.lineage_id
id_taxidLineage.head()

In [ ]:
# Root the lineage starting from the organism of interest
id_taxidLineage = str(organism_taxid) + id_taxidLineage.str.split(str(organism_taxid), expand=True)[1].astype(str)
id_taxidLineage.head()

In [ ]:
#id_rankorder = data.rank # rank is a function of dataframes
id_rankorder = data['rank']
id_rankorder.head()

In [ ]:
id_rankLineage = pd.Series()
for idx, lineage_list in id_taxidLineage.str.split("//").iteritems():
    rank_list = []
    for lin_id in lineage_list:
        lin_rank = id_rankorder[int(lin_id)]
        rank_list.append(lin_rank)
    id_rankLineage[str(idx)] = rank_list
id_rankLineage.head()

In [ ]:
rank_lin_df = id_rankLineage.to_frame(name = "rank_lineage")
rank_lin_df

In [ ]:
rank_lin_df = rank_lin_df.assign(rank_lineage = lambda x: x.rank_lineage.str.join("//"))
rank_lin_df.head()

In [ ]:
# Merge original dataframe to the new one with lineage rank
rank_lin_df.index = rank_lin_df.index.map(int)
df = data.join(rank_lin_df)
df.rename(columns = {"rank_lineage": "lineage_rank"}, inplace = True)
df = df[['name', 'rank', 'lineage_name', 'lineage_rank', 'lineage_id', 'sonof_id', 'sonof_name']] # reorder columns
df.head()

In [ ]:
#df.to_csv(path_or_buf = 'taxonomy_teuthida.csv', index_label = 'taxid')